[BEAM-5745] Fix annotation test for py3

HuangLED opened the pull request:
Root cause:  warnings.simplefilter method have limited scope in py3, which makes the "once" filter fail. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

HuangLED commented:
Run Python PostCommit

HuangLED commented:
R: @tvalentyn 

beamimport tap test passes for this PR.

HuangLED commented:
> It seems odd to set this globally once a function is called (and not in a context).

An alternative way, for the sake of fixing this failing test, is to explicitly call warnings.simplefilter() in each test case. This Option#2 of fix works as well.  The downside though, is that the discrepency between py2 and py3 would still exist. 

My hypothesis on the root cause, is that python2 and python3 treat unscoped global function call differently. (funny thing is, this warning.simplefilter is called in py3,  there is something else under the hood for which I don't know how) Didn't find a definitive documentation on this though, please fill in if you are aware of good discussions on this. (+python experts @pabloem  @tvalentyn )

Weighing these two options, I prefer option#1 in this PR because with this fix, this module by itself behave exactly the same behavior regardless in an py2 or py3 environment. And I don't think there is performance concern.  The downside here is the code does not look elegant enough. 

to wrap it up: 
Option#1:  Pro: this annotations.py module behaves the same in py2 and py3 ; Cons:  Code does not look elegant. 
Option#2:  Pro: only need to change test file.   Cons:  We will need to handle the py2/py3 discrepency in all related test cases, and will need to do that in the future non-testing module if anything depend on this annotations.py. 

Please make recommendations if you know an Option#3.  Thanks.

tvalentyn commented:
The original Python 2 implementation has issues to begin with... will add an extended comment in a bit.

tvalentyn commented:
1. The goal of `annotations.py` is to create a mechanism that will allow to print a warning message when some Beam APIs are used. It is well documented in the docstring https://github.com/apache/beam/blob/0acb027cd7d909732e5a07ed0d2cbbedf5fbc907/sdks/python/apache_beam/utils/annotations.py#L18
My intuition says we may be reinventing the wheel here, but without further research I don't have a recommendation at the moment.

2. One aspect of existing implementation is that we want such warnings be seen only once, if possible, to avoid noise. This is achieved by adding a warning filter, see: https://docs.python.org/3/library/warnings.html#warning-filter. Unfortunately, the restriction we impose in https://github.com/apache/beam/blob/0acb027cd7d909732e5a07ed0d2cbbedf5fbc907/sdks/python/apache_beam/utils/annotations.py#L94 is very broad, and will apply to **all** warnings which will be generated, not just warnings produced by annotated functions. This may have unintended side effects beyond the scope of the code the restriction is meant for. This is what @robertwb called out.
 
3. To understand why tests started to fail, we can add look into the filter list. Everytime we add a new warning filter, it is added to the beginning list, the order of filter application is defined in the documentation (see above). At any point, we can print current filter list with:
```
import pprint
pprint(warnings.filter)
```
By the time we run the annotated functions in `test_frequency`, the filter list on Python 2 looks like this: 
```
...
('always', None, <class 'numpy.lib.polynomial.RankWarning'>, None, 0),  # <--- Filter added by numpy library after our filter.                                                                                                                                          
('once', None, <type 'exceptions.Warning'>, None, 0),       # <--- Filter added by us. Notice that it applies to all warnings.
('ignore', None, <class 'urllib3.exceptions.DependencyWarning'>, None, 0),   # <--- Filters added before our filter.                                                                                                                                     
...
```

On Python 3, the filter list looks like this:

```
...
 ('default', None, <class 'Warning'>, None, 0),     # <--- !!! Something adds this rule.                                                                                                                                                          
... 
 ('always', None, <class 'numpy.lib.polynomial.RankWarning'>, None, 0),
 ('once', None, <class 'Warning'>, None, 0)       # <---- Rule we add
...
```
``` ('default', None, <class 'Warning'>, None, 0)``` overrides our behavior to print warnings only once. Therefore, the deprecation warnings are printed twice and the test fails. Might be interesting to track down what adds this rule. By setting `warnings.simplefilter("once")` inside `annotate` method, as this PR proposes, we basically start a fight with somebody over whose rule will be at the top.

I think we should reconsider whether we want to keep our annotation mechanism, if so, we need a better way to manage the warnings. Either not add any restrictions, or otherwise find a proper way to localize them. cc: @robertwb @aaltay

robertwb commented:
Thanks for the detailed analysis, @tvalentyn . I agree both with your assessment that we're just covering things up rather than finding the root cause, and that we should reconsider the whole annotations system. 

For the purpose of shipping a Python 3 compatible SDK, I propose that resolving this double-printing is low priority and we simply relax this test for Python 3 until all other kinks have been worked out.

tvalentyn commented:
For completeness, the overriding filter `('default', None, <class 'Warning'>, None, 0)` is added by unit test framework here:
https://github.com/python/cpython/blob/1a2b52bbd2c6513aaf5bb915890ea7647713d5df/Lib/unittest/runner.py#L161.

Full stacktrace:
```
  File "./setup.py", line 230, in <module>                                  
    'test': generate_protos_first(test),                           
  File "/tmp/py3env/lib/python3.5/site-packages/setuptools/__init__.py", line 140, in setup
    return distutils.core.setup(**attrs)                                      
  File "/usr/lib/python3.5/distutils/core.py", line 148, in setup                                                       
    dist.run_commands()                     
  File "/usr/lib/python3.5/distutils/dist.py", line 955, in run_commands
    self.run_command(cmd)                                                     
  File "/usr/lib/python3.5/distutils/dist.py", line 974, in run_command
    cmd_obj.run()                                                           
  File "./setup.py", line 165, in run
    super(cmd, self).run()
  File "/tmp/py3env/lib/python3.5/site-packages/setuptools/command/test.py", line 228, in run
    self.run_tests()
  File "/tmp/py3env/lib/python3.5/site-packages/setuptools/command/test.py", line 250, in run_tests
    exit=False,
  File "/usr/lib/python3.5/unittest/main.py", line 94, in __init__
    self.runTests()
  File "/usr/lib/python3.5/unittest/main.py", line 255, in runTests
    self.result = testRunner.run(self.test)
  File "/usr/lib/python3.5/unittest/runner.py", line 161, in run
    warnings.simplefilter(self.warnings)
  File ".../beam/sdks/python/apache_beam/utils/annotations.py", line 105, in my_simplefilter_patch
    traceback.print_stack(file=sys.stdout)    # <--- My monkey-patch.
```
So the behavior of annotations doesn't change on Python 3, but some internals of unit test framework did. I lean towards removing `warnings.simplefilter("once")` from our code, since it's not done right, and removing unit tests that  verify that behavior.

HuangLED commented:
Thanks Valentyn for sharing the research.   Agree that we'd better rethink the annotation machanism, which will require some dedicated efforts and beyond the scope of this PR. 

+1 to remove the line of simplefilter in py code. 

Regarding the unit test. I vote for keeping the unit test and adding simplefilter() in the test to counter act what is done by testing framework (a.k.a. the options#2 I mentioned earlier)   Thoughts?

HuangLED commented:
Run Python PostCommit

HuangLED commented:
Updated based on discussions: 
1) filter rule in annotation module is removed.  
2) Test case on the frequency limit is removed. 
3) In a few existing test case , added simplefilter to counter act whatever is added by testing framework.

tvalentyn commented on sdks/python/apache_beam/utils/annotations_test.py:
```diff
@@ -32,9 +32,10 @@ def test_deprecated_with_since_current_message(self):
       @deprecated(since='v.1', current='multiply', extra_message='Do this')
       def fnc_test_deprecated_with_since_current_message():
         return 'lol'
+      warnings.simplefilter("once")
```
Do we actually need this line? If the test does not pass with out it, most likely we need to fix the test.

tvalentyn commented on sdks/python/apache_beam/utils/annotations_test.py:
```diff
@@ -32,9 +32,12 @@ def test_deprecated_with_since_current_message(self):
       @deprecated(since='v.1', current='multiply', extra_message='Do this')
       def fnc_test_deprecated_with_since_current_message():
         return 'lol'
+      # In python2, the test framework by default adds an ignore filtering rule.
```
Thanks. Turns out DeprecationWarnings are ignored by default, not just on Python 2, see: https://docs.python.org/3/library/warnings.html. So I suggest the following wording (let's add a clarifying comment for all test methods that need this).
 
```
def test_deprecated_with_since_current_message(self):
  with warnings.catch_warnings(record=True) as w:
    # Deprecation warnings are ignored by default, turn them on to enable testing.
    warnings.simplefilter("once", DeprecationWarning)
```
Adding this to each individual test is the right thing, since we add a filter override inside `catch_warnings` context manager, so it will not affect other tests in the suite.

HuangLED replied:
Acked.  Thanks.

Though this PR is now in conflict with another non-trivial merge.  Does not seem trivial to get all resolved. 

Please let me know if there is any way better than copy over the master version and apply all my changes again.

tvalentyn commented:
One way to rebase your change on top of master, manually resolve merge conflict, and squash/fixup commits into 1 is via `git pull --rebase=interactive origin master`. Looking at the conflict, you'd have to resolve it manually or start over with the PR using a new branch, that has most recent code from master.

HuangLED commented:
Run Python PostCommit

tvalentyn commented:
Can you please rename the commit message? Looks like after squashing the commits we picked up the last commit's message, usually the first commit messages are more informative since they describe the purpose of the change. Thank you. The change LGTM as long as the test pass and you checked that they pass under Python 3 as well.

HuangLED commented:
Run Python PostCommit

HuangLED commented:
+ @aaltay for merge

HuangLED commented:
> Can you please rename the commit message? Looks like after squashing the commits we picked up the last commit's message, usually the first commit messages are more informative since they describe the purpose of the change. Thank you. The change LGTM as long as the test pass and you checked that they pass under Python 3 as well.

Done renaming the squashing commit.  

+ @swegner  for merge

udim commented:
@HuangLED @ihji @markflyhigh 
Please revert the change that removes `warnings.simplefilter("once")`, as it disables all deprecation notices for Beam SDK.
I've opened https://issues.apache.org/jira/browse/BEAM-5932 for this.
